internal/bits/bit_array.go

```diff
@@ -247,44 +248,61 @@ func (bA *BitArray) PickRandom() (int, bool) {
 	}
 
 	bA.mtx.Lock()
-	trueIndices := bA.getTrueIndices()
+	numTrueIndices := bA.getNumTrueIndices()
+	if numTrueIndices == 0 { // no bits set to true
+		bA.mtx.Unlock()
+		return 0, false
+	}
+	index := bA.getNthTrueIndex(cmtrand.Intn(numTrueIndices))
 	bA.mtx.Unlock()
-
-	if len(trueIndices) == 0 { // no bits set to true
+	if index == -1 {
 		return 0, false
 	}
+	return index, true
+}
 
-	return trueIndices[cmtrand.Intn(len(trueIndices))], true
+func (bA *BitArray) getNumTrueIndices() int {
+	count := 0
+	numElems := len(bA.Elems)
+	for i := 0; i < numElems; i++ {
+		count += bits.OnesCount64(bA.Elems[i])
+	}
+	return count
 }
 
-func (bA *BitArray) getTrueIndices() []int {
-	trueIndices := make([]int, 0, bA.Bits)
-	curBit := 0
+// getNthTrueIndex returns the index of the nth true bit in the bit array.
+// If there is no such value, it returns 0. It is required that the caller
+// ensures that n is less than or equal to the number of true bits in the bit array.
```

Comment:
This is requirement is not really a requirement, right? The code support this case, but returns -1.

Reply:
True!